app/src/main/java/uk/ac/ebi/atlas/search/JsonGeneSearchController.java

```diff
@@ -179,6 +137,68 @@ public String search(@RequestParam MultiValueMap<String, String> requestParams)
                         "checkboxFacetGroups", ImmutableList.of(MARKER_GENE.getTitle(), ORGANISM.getTitle())));
     }
 
+    @RequestMapping(value = "/json/search/marker-genes",
+            method = RequestMethod.GET,
+            produces = MediaType.APPLICATION_JSON_UTF8_VALUE)
+    public Boolean isMarkerGene(@RequestParam MultiValueMap<String, String> requestParams) {
+        if (isRequestParamsEmpty(requestParams)) {
+            return false;
+        }
+
+        GeneQuery geneQuery = geneIdSearchService.getGeneQueryByRequestParams(requestParams);
+        var geneIds = geneIdSearchService.search(geneQuery);
+
+        String emptyGeneIdError = geneIdEmptyValidation(geneIds);
+        if (emptyGeneIdError != null) {
+            return false;
+        }
+
+        List<Map.Entry<String, Map<String, List<String>>>> expressedGeneIdEntries =
+                getMarkerGeneProfileByGeneIds(geneIds);
+
+        var markerGeneFacets =
+                geneSearchService.getMarkerGeneProfile(
+                        expressedGeneIdEntries.stream()
+                                .map(Map.Entry::getKey)
+                                .toArray(String[]::new));
+
+        return markerGeneFacets != null && markerGeneFacets.size() > 0;
+    }
+
+    private List<Map.Entry<String, Map<String, List<String>>>> getMarkerGeneProfileByGeneIds(Optional<ImmutableSet<String>> geneIds) {
+        // We found expressed gene IDs, let’s get to it now...
+        var geneIds2ExperimentAndCellIds =
+                geneSearchService.getCellIdsInExperiments(geneIds.get().toArray(new String[0]));
+
+        return geneIds2ExperimentAndCellIds.entrySet().stream()
+                        .filter(entry -> !entry.getValue().isEmpty())
+                        .collect(toList());
+    }
+
+    private boolean isRequestParamsEmpty(MultiValueMap<String, String> requestParams) {
+        return requestParams == null
+                || requestParams.size() == 0
+                || (requestParams.containsKey("q") && Objects.equals(requestParams.getFirst("q"), ""));
```

Comment:
Won’t this violate the contract in the comment?
```java
        // We support currently only one query term; in the unlikely case that somebody fabricates a URL with more than
        // one we’ll build the query with the first match.
```

Meaning that something like `?q=&q=some+valid+query` would fail, when it should search for "some valid query". Maybe we could add a test to describe the expected behaviour?

Reply:
Refactored the code and add a couple of more tests to some of the classes.